Backend/FcTypes.hs

```diff
@@ -465,3 +465,55 @@ instance PrettyPrint (FcProgram a) where
   ppr (FcPgmValDecl  valbind  pgm) = ppr valbind  $$ ppr pgm
   ppr (FcPgmTerm tm)               = ppr tm
   needsParens _ = False
+
+-- * Term Size
+-- -----------
+
+-- We define size as: number of nodes in abstract syntax tree, for terms and types
+class Size a where
+  size :: a -> Int
+
+instance Size a => Size [a] where
+  size = foldr ((+) . size) 0
+
+instance Size (FcProgram a) where
+  size (FcPgmDataDecl _ pgm) = size pgm
+  size (FcPgmValDecl  _ pgm) = size pgm
+  size (FcPgmTerm     tm   ) = size tm
+
+instance Size (FcType) where
+  size (FcTyVar     _x ) = 1
+  size (FcTyCon     _c ) = 1
+  size (FcTyAbs _x  ty ) = size ty + 1
+  size (FcTyApp ty1 ty2) = size ty1 + size ty2
+
+instance Size (FcTerm a) where
+  size (FcTmVar     _x   ) = 1
+  size (FcTmDataCon _c   ) = 1
+  size (FcTmERROR   _e ty) = size ty + 1
+
+  size (FcTmAbs   _x ty t    ) = size ty + size t + 1
+  size (FcTmTyAbs _x t       ) = size t + 1
+  size (FcTmLet   _x ty t1 t2) = size ty + size t1 + size t2 + 1
+
+  size (FcTmTyApp    t  ty) = size ty + size t
+  size (FcTmApp      t1 t2) = size t1 + size t2
+
+  size (FcTmCaseFc         t as) = size t + size as + 1
+  size (FcTmCaseTc ty1 ty2 t as) = size ty1 + size ty2 + size t + size as + 1
+
+instance Size (FcAlt a) where
+  size (FcAltTc p gs) = size p + size gs + 1
+  size (FcAltFc p t ) = size p + size t + 1
```

Comment:
```suggestion
  size (FcAltTc p gs) = size p + size gs
  size (FcAltFc p t ) = size p + size t
```
Personally I'd remove the `+1`s here. A good guideline is that every case expression has alternatives in it, and every alternative contains some patterns and a right-hand side. The `FcTmCaseTc` and `FcTmCaseFc` cases above have already done their `+1` so it should be OK.